From 6a7aa0c718d69c437532dc83a6f111bc61fe40a1 Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Tue, 13 Oct 2026 20:56:58 +0200
Subject: [PATCH] real_type: the mfront interface of the material laws accepted

The behaviours using material laws (@MaterialLaw) have them generated by
the mfront interface, with the DSL options given on the command line,
including real_type: the behaviours of the examples in float were refused.
This interface generates C++ functions computed in double, called by the
behaviours with conversions: it accepts any numerical type.
---
 mfront/include/MFront/DSLBase.hxx |  3 ++-
 mfront/src/DSLBase.cxx            | 12 ++++++++----
 2 files changed, 10 insertions(+), 5 deletions(-)

diff --git a/mfront/include/MFront/DSLBase.hxx b/mfront/include/MFront/DSLBase.hxx
index 0df125e10..aea0b1c7d 100644
--- a/mfront/include/MFront/DSLBase.hxx
+++ b/mfront/include/MFront/DSLBase.hxx
@@ -199,7 +199,8 @@ namespace mfront {
      * \brief check that the numerical type selected by the `real_type`
      * option is supported by the given interfaces: a type other than
      * `double` is only supported by the `generic` and `generic-parallel`
-     * interfaces
+     * interfaces, and by the `mfront` interface of the material laws used
+     * by the behaviours (computed in double)
      * \param[in] d: material knowledge description
      * \param[in] interfaces: names of the interfaces
      */
diff --git a/mfront/src/DSLBase.cxx b/mfront/src/DSLBase.cxx
index 6259aa66a..e5be7758f 100644
--- a/mfront/src/DSLBase.cxx
+++ b/mfront/src/DSLBase.cxx
@@ -168,12 +168,15 @@ namespace mfront {
     if (t == "double") {
       return;
     }
+    // the mfront interface generates the C++ functions of the material laws
+    // used by the behaviours (@MaterialLaw), computed in double and called
+    // with conversions
     for (const auto& i : interfaces) {
-      if ((i != "generic") && (i != "generic-parallel")) {
+      if ((i != "generic") && (i != "generic-parallel") && (i != "mfront")) {
         tfel::raise("DSLBase::checkRealTypeSupport: the numerical type '" + t +
                     "' (real_type option) is not supported by interface '" + i +
-                    "', only by the 'generic' and 'generic-parallel' "
-                    "interfaces");
+                    "', only by the 'generic', 'generic-parallel' and "
+                    "'mfront' interfaces");
       }
     }
   }  // end of checkRealTypeSupport
@@ -229,7 +232,8 @@ namespace mfront {
             {DSLBase::realTypeOption,
              "string specifying the numerical type of the `generic` and "
              "`generic-parallel` interfaces, `double` (default) or `float`. "
-             "The other interfaces and the models require `double`"},
+             "The other interfaces (except the `mfront` interface of the "
+             "material laws) and the models require `double`"},
             {DSLBase::overridingParameters,
              "map allowing to override material properties, parameters and "
              "external state variables by parameters"}};
-- 
2.43.0

